File: Java/src/uk/ac/ucl/hvtp/client/HyperverseClient.java
package uk.ac.ucl.hvtp.client;

import uk.ac.ucl.hvtp.OnPacketCallback;
import uk.ac.ucl.hvtp.Packet;
import uk.ac.ucl.hvtp.PacketHeader;
import uk.ac.ucl.hvtp.PacketSerializer;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.concurrent.atomic.AtomicLong;

public class HyperverseClient implements Runnable, IHyperverseClient {
	private static final AtomicLong COUNTER = new AtomicLong();

	private final Socket socket;

	private volatile boolean running = true;

	private Thread thread;
	private OnPacketCallback callback;

	public HyperverseClient(Socket socket) {
		this.socket = socket;
	}

	@Override
	public void create(OnPacketCallback callback) {
		if (running) {
			return;
		}

		running = true;

		this.callback = callback;

		thread = new Thread(this);
		thread.setName("Hyperverse Client #" + COUNTER.incrementAndGet());
		thread.start();
	}

	@Override
	public void run() {
		while (running) {
			try {
				Packet packet = receivePacket();
				callback.onPacketReceived(packet);
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	private Packet receivePacket() throws IOException {
		InputStream inputStream = socket.getInputStream();

		byte[] bytes = inputStream.readNBytes(PacketHeader.SIZE_IN_BYTES);

		PacketHeader header = PacketSerializer.deserializeHeader(bytes);

		byte[] payload = inputStream.readNBytes(header.getLength());

		return new Packet(header, payload);
	}

	public synchronized void sendPacket(Packet packet) throws IOException {
		byte[] bytes = PacketSerializer.serialize(packet);

//		out.write(bytes);
		OutputStream outputStream = socket.getOutputStream();
		outputStream.write(bytes);
	}

	@Override
	public void close() throws IOException {
		if (!running) {
			return;
		}

		running = false;

		try {
			thread.join(1000);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}

		socket.close();
	}
}

File: Unity/Hyperverse Client/Assets/HyperverseClient.cs
using System.Net.Sockets;
using System.Threading;
using UnityEngine;
using HVTP;

namespace HVTP
{
    public class HyperverseClient
    {
        private volatile bool running = false;

        private Thread thread;
        private Socket socket;
        private OnPacketReceived callbackDelegate;

        public delegate void OnPacketReceived(Packet packet);

        // public delegate void OnSceneGraphReceived(byte[] bytes);

        public void Create(string ip, int port, OnPacketReceived callbackDelegate)
        {
            if (running)
            {
                return;
            }

            socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            socket.Connect(ip, port);

            if (!socket.Connected)
            {
                Debug.Log("Connection failed!");
                return;
            }

            running = true;

            this.callbackDelegate = callbackDelegate;

            // Upon connection, the Hyperverse server will pass us the entire scene-graph under the INIT packet.
            // The client update function will run on a separate thread. 

            thread = new Thread(Run);
            thread.Start();
        }

        public void SetCallbackDelegate(OnPacketReceived callbackDelegate) {
            this.callbackDelegate = callbackDelegate;
        }

        private Packet ReceivePacket()
        {
            byte[] bytes = new byte[PacketHeader.SIZE_IN_BYTES];

            socket.Receive(bytes, 0, bytes.Length, SocketFlags.None);

            PacketHeader header = PacketSerializer.deserializeHeader(bytes);

            byte[] payload = new byte[header.GetLength()];

            socket.Receive(payload, 0, payload.Length, SocketFlags.None);

            return new Packet(header, payload);
        }

        public void SendPacket(Packet packet)
        {
            byte[] bytes = PacketSerializer.serialize(packet);
            
            socket.Send(bytes, bytes.Length, SocketFlags.None);   
        }

        public void Destroy()
        {
            if (!running)
            {
                return;
            }

            running = false;
            //thread.Join();
            thread.Abort();
            socket.Dispose();
        }

        public void Run()
        {
            while (running)
            {
                // ReceivePacket will block here until another message is passed.

                Packet packet = ReceivePacket();
                callbackDelegate(packet);
            }
        }
    }
}